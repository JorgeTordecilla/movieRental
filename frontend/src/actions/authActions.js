import Swal from "sweetalert2";
import { fetchConToken, fetchSinToken } from "../helpers/fetch";
import { types } from "../types/types";

export const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await fetchSinToken("auth", { email, password }, "POST");
    const body = await resp.json();
    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(
        login({
          uid: body.uid,
          name: body.name,
          balance: body.balance,
          isPremium: body.isPremium,
        })
      );
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    const resp = await fetchSinToken(
      "auth/new",
      { name, email, password },
      "POST"
    );
    const body = await resp.json();
    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        login({
          uid: body.uid,
          name: body.name,
          balance: body.balance,
          isPremium: body.isPremium,
        })
      );
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchConToken("auth/renew");
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        login({
          uid: body.uid,
          name: body.name,
          balance: body.balance,
          isPremium: body.isPremium,
        })
      );
    } else {
      dispatch(checkingFinish());
    }
  };
};

export const checkingFinish = () => ({ type: types.authCheckingFinish });

export const startLogout = () => {
  return (dispatch) => {
    localStorage.removeItem("token");
    dispatch(logout());
  };
};

export const startRemoveUser = (uid) => {
  return async (dispatch) => {
    await fetchConToken("users/remove", { uid }, "POST");
    localStorage.removeItem("token");
    dispatch(logout());
  };
};

const logout = () => ({ type: types.authLogout });

export const startRenting = (uid, balance, isPremium, moviePrice) => {
  return async (dispatch) => {
    const resp = await fetchConToken(
      "users/rent",
      { uid, balance, isPremium, moviePrice },
      "POST"
    );
    const body = await resp.json();
    dispatch(startChecking());
    if (body.ok) {
      console.log(body);
    } else {
      dispatch(checkingFinish());
    }
  };
};
