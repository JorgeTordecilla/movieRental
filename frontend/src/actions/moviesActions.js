import { fetchConToken } from "../helpers/fetch";
import { types } from "../types/types";

export const moviesLoaded = (movies) => ({
  type: types.moviesLoaded,
  payload: movies,
});

export const rentMoviesLoaded = (movies) => ({
  type: types.moviesLoadedRent,
  payload: movies,
});

export const moviesRent = (movies) => ({
  type: types.moviesRent,
  payload: movies,
});

export const moviesStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchConToken("movies/movies");
      const body = await resp.json();
      dispatch(moviesLoaded(body.movies));
    } catch (error) {
      console.log(error);
    }
  };
};

export const moviesStartRent = (movieUid, userUid) => {
  return async (dispatch) => {
    try {
      await fetchConToken("movies/rent", { movieUid, userUid }, "POST");
      dispatch(moviesStartLoading());
    } catch (error) {
      console.log(error);
    }
  };
};

export const rentMoviesStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchConToken("movies/rentmovies");
      const body = await resp.json();
      dispatch(rentMoviesLoaded(body.movies));
    } catch (error) {
      console.log(error);
    }
  };
};

export const returnRentMoviesStartLoading = (uid, movieUid) => {
  return async (dispatch) => {
    try {
      await fetchConToken("movies/returnmovies", { uid, movieUid }, "POST");

      dispatch(rentMoviesStartLoading());
    } catch (error) {
      console.log(error);
    }
  };
};
