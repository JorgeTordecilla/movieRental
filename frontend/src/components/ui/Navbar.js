import React from "react";
import { Link, NavLink } from "react-router-dom";

import { useDispatch, useSelector } from "react-redux";
import { startLogout } from "../../actions/authActions";

export const Navbar = () => {
  const dispatch = useDispatch();

  const { name, balance } = useSelector((state) => state.auth);

  const handleLogout = () => {
    dispatch(startLogout());
  };

  return (
    <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
      <div className="container-fluid">
        <Link className="navbar-brand" to="/">
          MoviesRental
        </Link>

        <button
          className="navbar-toggler"
          type="button"
          data-bs-toggle="collapse"
          data-bs-target="#navbarText"
          aria-controls="navbarText"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span className="navbar-toggler-icon"></span>
        </button>

        <div className="collapse navbar-collapse" id="navbarText">
          <ul className="navbar-nav me-auto mb-2 mb-lg-0">
            <li className="nav-item">
              <NavLink
                className={({ isActive }) =>
                  "nav-item nav-link " + (isActive ? "active" : "")
                }
                to="/user"
              >
                Profile
              </NavLink>
            </li>
          </ul>
          <span className="navbar-text">
            <div className="collapse navbar-collapse w-100 order-3 dual-collapse2 d-flex justify-content-end">
              <ul className="navbar-nav ml-auto">
                <span className="nav-item nav-link text-info">{name}</span>
                <span className="nav-item nav-link text-info">${balance}</span>
                <button
                  className="nav-item nav-link btn"
                  onClick={handleLogout}
                >
                  Logout
                </button>
              </ul>
            </div>
          </span>
        </div>
      </div>
    </nav>
  );
};
