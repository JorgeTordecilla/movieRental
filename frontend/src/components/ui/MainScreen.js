import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { moviesLoaded, moviesStartLoading } from "../../actions/moviesActions";
import { Navbar } from "./Navbar";
import { MoviesList } from "../movies/MoviesList";

export const MainScreen = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    (async () => {
      await dispatch(moviesStartLoading());
    })();
  }, [dispatch]);

  const { movies } = useSelector((state) => state.movies);

  const handleSortAsc = () => {
    movies.sort(function (a, b) {
      return a.genre.localeCompare(b.genre);
    });
    dispatch(moviesLoaded(movies));
  };

  const handleSortDsc = () => {
    movies.sort(function (a, b) {
      return b.genre.localeCompare(a.genre);
    });
    dispatch(moviesLoaded(movies));
  };
  return (
    <>
      <Navbar />
      <div className="container">
        <button className="btn btn-info mt-2 " onClick={handleSortAsc}>
          ↑
        </button>
        <button className="btn btn-info mt-2 ml-2 " onClick={handleSortDsc}>
          ↓
        </button>
        <MoviesList movies={movies} />
      </div>
    </>
  );
};
