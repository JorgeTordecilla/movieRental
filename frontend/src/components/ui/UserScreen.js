import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { rentMoviesStartLoading } from "../../actions/moviesActions";
import { startRemoveUser } from "../../actions/authActions";
import { Navbar } from "./Navbar";
import { MoviesRentList } from "../movies/MoviesRentList";
import Swal from "sweetalert2";

export const USerScreen = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    (async () => {
      await dispatch(rentMoviesStartLoading());
    })();
  }, [dispatch]);

  const { rentals } = useSelector((state) => state.movies);
  const auth = useSelector((state) => state.auth);

  let userMod = { ...auth };
  delete userMod.checking;
  delete userMod.uid;

  userMod = Object.entries(userMod);
  userMod.push(["password", null]);
  userMod.push(["email", null]);

  const handleDelteUser = () => {
    Swal.fire({
      title: "Do you want to delete your user?",
      showDenyButton: true,
      showCancelButton: true,
      confirmButtonText: "Yes",
      denyButtonText: "No",
      customClass: {
        actions: "my-actions",
        cancelButton: "order-1 right-gap",
        confirmButton: "order-2",
        denyButton: "order-3",
      },
    }).then((result) => {
      if (result.isConfirmed) {
        Swal.fire("Saved!", "", "success");
        dispatch(startRemoveUser(auth.uid));
      } else if (result.isDenied) {
        Swal.fire("Changes are not saved", "", "info");
      }
    });
  };

  return (
    <>
      <Navbar />
      {rentals &&
        rentals.map((rental) => (
          <MoviesRentList key={rental.date} {...rental} />
        ))}
      <div className="container">
        <h1>User</h1>
        {userMod.map((x) => (
          <div className="form-floating" key={x}>
            <textarea className="form-control" id="floatingTextarea"></textarea>
            <label htmlFor="floatingTextarea">{x[0]}</label>
          </div>
        ))}
        <div className="btn-group-sm   d-flex align-content-center">
          <button className="btn btn-outline-info">Actulizar</button>
          <button
            className="btn btn-outline-danger ml-2"
            onClick={handleDelteUser}
          >
            Dar de baja
          </button>
        </div>
      </div>
    </>
  );
};
