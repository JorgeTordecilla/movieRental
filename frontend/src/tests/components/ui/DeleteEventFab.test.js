import { mount } from "enzyme";
import { Provider } from "react-redux";
import configuerStore from "redux-mock-store";
import thunk from "redux-thunk";
import { eventStartDelete } from "../../../actions/eventsActions";
import { DeleteEventFab } from "../../../components/ui/DeleteEventFab";
jest.mock("../../../actions/eventsActions", () => ({
  eventStartDelete: jest.fn(),
}));
const initState = {};
const middlewares = [thunk];
const mockStore = configuerStore(middlewares);
const store = mockStore(initState);
store.dispatch = jest.fn();
const { dispatch, clearActions } = store;

const wrapper = mount(
  <Provider store={store}>
    <DeleteEventFab />
  </Provider>
);
describe("pruebas en <DeleteEventFab/>", () => {
  beforeEach(() => {
    clearActions();
  });
  test("debe de hacer match con el snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("debe llamar la accion eventStartDelete(", () => {
    wrapper.find("button").simulate("click");
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(eventStartDelete).toHaveBeenCalled();
  });
});
