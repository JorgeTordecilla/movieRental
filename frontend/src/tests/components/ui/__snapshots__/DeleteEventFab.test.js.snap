// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <DeleteEventFab/> debe de hacer match con el snapshot 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    aria-hidden="true"
    className="fa fa-trash"
  />
  <span>
     Borrar Evento
  </span>
</button>
`;
